Security Deployment Guidelines Tutorial (#3318)
ros2/ros2_documentation: source/Tutorials/Advanced/Security/Deployment-Guidelines.rst

Deployment Guidelines
=====================

**Goal:** Understand the best practices when deploying security artifacts into production systems.

**Tutorial level:** Advanced

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

Typical deployment scenarios often involve shipping containerized applications, or packages, into remote systems.
Special attention should be payed when deploying security enabled applications, requiring users to reason about the sensitivity of packaged files.

Complying with the `DDS Security standard <https://www.omg.org/spec/DDS-SECURITY/1.1/About-DDS-SECURITY/>`_,
the ``sros2`` package provides a collection of utilities for managing security under ROS 2 environments in a highly modular and flexible fashion.

Basic core guidelines on how to organize the different certificates, keys and directories remains a critical factor to avoid compromising the security of the system.
This includes protection-awareness and criteria for selecting the minimum set of necessary files to be deployed upon remote production systems for minimizing security exposure.

Prerequisites
-------------

> * A docker installation with the compose plugin. Please refer to the installation steps detailed in `Docker installation <https://docs.docker.com/engine/install/>`_ and `Compose Plugin <https://docs.docker.com/compose/install>`_.

Reviewer: ```suggestion
* A docker installation with the compose plugin.
Please refer to the installation steps detailed in `Docker installation <https://docs.docker.com/engine/install/>`_ and `Compose Plugin <https://docs.docker.com/compose/install>`_.
```

Author: (I am leaving two white spaces at the beginning of the second sentence just to be in the same enumeration item)

  Required ``Security`` environment variables are sourced as well as the necessary keystore files from the shared volume.

The compose configuration yaml file can be downloaded with:

.. code-block:: bash

  # Download the compose file
  wget https://raw.githubusercontent.com/ros2/ros2_documentation/{DISTRO}/source/Tutorials/Advanced/Security/resources/deployment_gd/compose.deployment.yaml

Running the example
-------------------

In the same working directory ``~/security_gd_tutorial``, run:

.. code-block:: bash

  # Start the example
  docker compose -f compose.deployment.yaml up

This should result in the following output:

- *tutorial-listener-1*: ``Found security directory: /keystore/enclaves/talker_listener/listener``
- *tutorial-talker-1*: ``Found security directory: /keystore/enclaves/talker_listener/talker``
- *tutorial-listener-1*: ``Publishing: 'Hello World: <number>'``
- *tutorial-talker-1*: ``I heard: [Hello World: <number>]``

Examining the containers
^^^^^^^^^^^^^^^^^^^^^^^^

While having the containers running that simulate the two remote devices for this tutorial, attach to each of them by opening two different terminals and enter:

.. code-block:: bash

>   #Terminal 1

Reviewer: ```suggestion
  # Terminal 1
```

  docker exec -it tutorial-listener-1 bash
  cd keystore
  tree

>   #Terminal 2

Reviewer: ```suggestion
  # Terminal 2
```

  docker exec -it tutorial-talker-1 bash
  cd keystore
  tree

A similar output to the one depicted below should be obtained:

.. code-block:: bash

>   #Terminal 1

Reviewer: ```suggestion
  # Terminal 1
```

  keystore
   ├── enclaves
   │   ├── governance.p7s
   │   ├── governance.xml
   │   └── talker_listener
   │       └── listener
   │           ├── cert.pem
   │           ├── governance.p7s
   │           ├── identity_ca.cert.pem
   │           ├── key.pem
   │           ├── permissions_ca.cert.pem
   │           ├── permissions.p7s
   │           └── permissions.xml
   └── public
       ├── ca.cert.pem
       ├── identity_ca.cert.pem
       └── permissions_ca.cert.pem

>   #Terminal 2

Reviewer: ```suggestion
  # Terminal 2
```

| private                |       ✓      |       ✕       |         High        |
+------------------------+--------------+---------------+---------------------+
| enclaves               |       ✓      |       ✓       |        Medium       |
+------------------------+--------------+---------------+---------------------+

Building a deployment scenario
------------------------------

To illustrate a simple deployment scenario, a new docker image will be built on top of the one provided by ``ros:<DISTRO>``.
Starting from the image, three containers will be created with the aim of:

* Initializing the keystore in a local host's shared volume.
* Simulating two deployed remote devices that interact with each other in a secure way.

In this example, the local host serves as the organization's system.
Let us start by creating a workspace folder:

.. code-block:: bash

  mkdir ~/security_gd_tutorial
  cd ~/security_gd_tutorial

Generating the Docker Image
^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order to build a new docker image, a Dockerfile is required.
The one proposed for this tutorial can be retrieved with the following command:

.. code-block:: bash

  # Download the Dockerfile

>   wget https://raw.githubusercontent.com/ros2/ros2_documentation/rolling/source/Tutorials/Advanced/Security/resources/deployment_gd/Dockerfile

Reviewer 2: ```suggestion
  wget https://raw.githubusercontent.com/ros2/ros2_documentation/{DISTRO}/source/Tutorials/Advanced/Security/resources/deployment_gd/Dockerfile
```

Author: I applied both of them, but I was doubting if the tutorial would be back ported to previous versions, so I dont know if the files would be available in those cases

Reviewer 2: if this PR is not backported, this tutorial is not either available for backward distribution. resolving anyway.

Now, build the docker image with the command:

.. code-block:: bash

  # Build the base image
  docker build -t ros2_security/deployment_tutorial .

Understanding the compose file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A compose configration file takes an image to create containers as services.
In this tutorial, three services are defined within the configuration:

* *keystore-creator*: That, similarly to previous tutorials, it internally initializes a new keystore tree directory.
  This will create *enclaves/* *public/* and *private/*, which are explained in more detail in `ROS 2 Security enclaves <https://design.ros2.org/articles/ros2_security_enclaves.html>`_.
  The ``keystore`` directory is configured to be a shared volume across containers.

* *listener* and *talker*: Act as the remote device actors in this tutorial.
  Required ``Security`` environment variables are sourced as well as the necessary keystore files from the shared volume.

The compose configuration yaml file can be downloaded with:

.. code-block:: bash

  # Download the compose file

>   wget https://raw.githubusercontent.com/ros2/ros2_documentation/rolling/source/Tutorials/Advanced/Security/resources/deployment_gd/compose.deployment.yaml

Reviewer 2: ```suggestion
  wget https://raw.githubusercontent.com/ros2/ros2_documentation/{DISTRO}/source/Tutorials/Advanced/Security/resources/deployment_gd/compose.deployment.yaml
```

* Granting read-only permissions to the ``enclaves/`` directory contents.
* If a PKCS#11 compliant URI is given for generating enclave's private keys, a `Hardware Security Module (HSM) <https://en.wikipedia.org/wiki/Hardware_security_module>`_ could be used to store them.

The following table depicts a summary of the previous statements relating the Keystore directory with the Recommended location:

+------------------------+--------------+---------------+---------------------+
| Directory / Location   | Organization | Target Device | Material Sensitivity|
+========================+==============+===============+=====================+
| public                 |       ✓      |       ✓       |         Low         |
+------------------------+--------------+---------------+---------------------+
| private                |       ✓      |       ✕       |         High        |
+------------------------+--------------+---------------+---------------------+
| enclaves               |       ✓      |       ✓       |        Medium       |
+------------------------+--------------+---------------+---------------------+

Building a deployment scenario
------------------------------

> To illustrate a simple deployment scenario, a docker image will be built which is intented to be the remote production target device actor.

Reviewer: ```suggestion
To illustrate a simple deployment scenario, a docker image will be built which is intended to be the remote production target device actor.
```

> To illustrate a simple deployment scenario, a docker image will be built which is intented to be the remote production target device actor.

Reviewer 2: Since this is tutorial, i would think it would be nice if tutorial is the common use case or user likely do so. having docker container in production would not be really common in the edge device? probably we would use container runtime low level implementation runC or so on, but using docker is also security breach since container likely has root permission. Would it be nice if we have this tutorial with physical edge device?

Author: The aim of using a docker is just to simply recreate another virtual or "edge" physical device within the local host without needing another extra system for running the tutorial. In addition, it is not that rare to find virtualized applications in the deployed production systems. In the case of having a physical device both guidelines and and usage of security artifacts can also be applied in a similar fashion.

Regarding Docker, root and permissions, in the tutorial the user is encouraged to keep security material under specific permissions and/or within other restricted users.

Reviewer 2: i understand what you mean here to explain how to create and put the security files virtually. i guess that is not a hard requirement. I was thinking that security file access permission are important with this guideline. so just wondering that this operation guideline is really secured?

> it is not that rare to find virtualized applications in the deployed production systems

if that is the case here, probably using `root` permission is not really good idea? because that is user tend to do?

> Perhaps you could have the user create the keystore via commands, then run a compose file that would clearly mount only the volumes/paths as required for each container in the talker and listener example from the above README.

👍 i think this makes sense to me.

> In this example, localhost will serve as the organization's system.

Reviewer: ```suggestion
In this example, the local host will serve as the organization's system.
```

To test security capabilities, a secure listener will be launched on the remote system, whereas a secure talker will be launched on the local host.
Let us start by creating a workspace tree with the following sub-folders:

.. code-block:: bash

  mkdir -p ~/deploy_gd_tutorial/remote_system
  mkdir ~/deploy_gd_tutorial/keystore

Generating a Keystore and necessary Enclaves
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Similarly to previous tutorials, intialize a new keystore tree directory.

Reviewer: ```suggestion
Similarly to previous tutorials, initialize a new keystore tree directory.
```

This will create *enclaves/* *public/* and *private/* directories, which are explained in more detail in `ROS 2 Security enclaves <https://design.ros2.org/articles/ros2_security_enclaves.html>`_.

.. code-block:: bash

  # Source ROS installation
  source /opt/ros/${ROS_DISTRO}/setup.bash
  # Initialize a new keystore directory
  ros2 security create_keystore ~/deploy_gd_tutorial/keystore

Next, create an enclave for the local talker node within the */keystore* directory.

.. code-block:: bash

  # Create secure talker's enclave
  ros2 security create_enclave ~/deploy_gd_tutorial/keystore /talker_listener/talker

At this point, step into the remote_system workspace, create the corresponding enclave and copy just the *public/* and *enclaves/* directories to the current one.
Those security artifacts will be needed by the remote system to enable listener's security.
For the sake of simplicity, the same CA is used within this enclave for both identity and permissions.

> Note that *private/* folder is not moved but left in localhost (organization).

Reviewer: ```suggestion
Note that the *private/* folder is not moved but left in the local host (organization).
```

Deployment Guidelines
=====================

**Goal:** Understand the best practices when deploying security artifacts into production systems.

**Tutorial level:** Advanced

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

> Typical deployment scenarios oftenly involve shipping containerized applications, or packages, into remote systems.

Reviewer: ```suggestion
Typical deployment scenarios often involve shipping containerized applications, or packages, into remote systems.
```

Special attention should be payed when deploying security enabled applications, requiring users to reason about the sensitivity of packaged files.

Complying with the `DDS Security standard <https://www.omg.org/spec/DDS-SECURITY/1.1/About-DDS-SECURITY/>`_,
the ``sros2`` package provides a collection of utilities for managing security under ROS 2 environments in a highly modular and flexible fashion.

Basic core guidelines on how to organize the different certificates, keys and directories remains a critical factor to avoid compromising the security of the system.
This includes protection-awareness and criteria for selecting the minimum set of necessary files to be deployed upon remote production systems for minimizing security exposure.

Prerequisites
-------------

* A docker installation available. Please refer to the installation steps detailed in `Docker installation <https://docs.docker.com/engine/install/>`_.

> * (Recommended) A basic understanding on `ROS2 Security design <https://design.ros2.org/articles/ros2_dds_security.html>`_.

Reviewer: ```suggestion
* (Recommended) A basic understanding on `ROS 2 Security design <https://design.ros2.org/articles/ros2_dds_security.html>`_.
```

    * :doc:`The-Keystore`
    * :doc:`Access-Controls`

General Guidelines
------------------

ROS 2 leverages DDS Security extensions to ensure security on message exchanges within the same enclave.
The different signed files and certificates within an enclave are generated from the private keys and certificates of a `Certificate Authority (CA) <https://en.wikipedia.org/wiki/Certificate_authority>`_ trusted entity.
In fact, two different CA's can be selected for identity and permissions, per enclave.
Those CA artifacts are stored inside ``private/`` and ``public/`` sub-directories of a `Keystore <https://design.ros2.org/articles/ros2_security_enclaves.html>`_ with the following folder structure:

.. code-block:: text

  keystore
  ├── enclaves
  │   └── ...
  │       └── ...
  ├── private
  │   └── ...
  └── public
      └── ...

A good practice for the creation and usage of a certain Certificate Authority on a typical deployment for a production system, is to:

#. Create it within the organization system intended for internal use only.
#. Generate/modify desired enclaves bearing in mind that:

    A. Not all the generated enclaves should be embarked into all target devices.

>     #. A reasonable way to proceed would be having one enclave per application, allowing for a separation of concerns.

Reviewer: ```suggestion
    * Not all the generated enclaves should be deployed to all target devices.
    * A reasonable way to proceed would be having one enclave per application, allowing for a separation of concerns.
```

+------------------------+--------------+---------------+---------------------+
| private                |       ✓      |       ✕       |         High        |
+------------------------+--------------+---------------+---------------------+
| enclaves               |       ✓      |       ✓       |        Medium       |
+------------------------+--------------+---------------+---------------------+

Building a deployment scenario
------------------------------

To illustrate a simple deployment scenario, a docker image will be built which is intented to be the remote production target device actor.
In this example, localhost will serve as the organization's system.
To test security capabilities, a secure listener will be launched on the remote system, whereas a secure talker will be launched on the local host.
Let us start by creating a workspace tree with the following sub-folders:

.. code-block:: bash

  mkdir -p ~/deploy_gd_tutorial/remote_system
  mkdir ~/deploy_gd_tutorial/keystore

Generating a Keystore and necessary Enclaves
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Similarly to previous tutorials, intialize a new keystore tree directory.

> This will create *enclaves/* *public/* and *private/* directories, which are explained in more detail in `ROS2 Security enclaves <https://design.ros2.org/articles/ros2_security_enclaves.html>`_.

Reviewer: ```suggestion
This will create *enclaves/* *public/* and *private/* directories, which are explained in more detail in `ROS 2 Security enclaves <https://design.ros2.org/articles/ros2_security_enclaves.html>`_.
```

.. code-block:: text

  enclaves
  ├── governance.p7s
  ├── governance.xml
  └── talker_listener
      ├── listener
      │   ├── cert.pem
      │   ├── governance.p7s
      │   ├── identity_ca.cert.pem
      │   ├── key.pem
      │   ├── permissions_ca.cert.pem
      │   ├── permissions.p7s
      │   └── permissions.xml
      └── talker
          ├── cert.pem
          ├── governance.p7s
          ├── identity_ca.cert.pem
          ├── key.pem
          ├── permissions_ca.cert.pem
          ├── permissions.p7s
          └── permissions.xml

Now, create and populate the */keystore* directory that will be copied onto the remote system with only the necessary files.

.. code-block:: bash

  # Move to remote system path
  cd ~/deploy_gd_tutorial/remote_system
  # Ship governance files, listener enclave and public/ directories only
  # minimizing security threat
  mkdir -p keystore/enclaves/talker_listener
  cp -R ../keystore/public keystore
  cp -R ../keystore/enclaves/governance.* keystore/enclaves
  cp -R ../keystore/enclaves/talker_listener/listener keystore/enclaves/talker_listener

> After the former commands, the current ``~/deploy_gd_tutorial/remote_system`` directory should be:

Reviewer: ```suggestion
After the previous commands, the current ``~/deploy_gd_tutorial/remote_system`` directory should be:
```

  COPY entrypoint.sh /entrypoint.sh
  RUN chmod +x /entrypoint.sh

  ENTRYPOINT ["/entrypoint.sh"]

The resultant directory structure should be the one depicted below:

.. code-block:: text

  remote_system
    ├── Dockerfile
    ├── entrypoint.sh
    └── keystore
        ├── enclaves
        │   ├── ...
        └── public
            ├── ...

Build the docker image with the command:

.. code-block:: bash

  # Build remote's system image
  docker build -t ros2_security/deployment_tutorial .

Running the example
-------------------

Launch the following commands in two different terminals:

Open a new terminal and run:

.. code-block:: bash

    # Start remote system container
    docker run -it ros2_security/deployment_tutorial

Then, open a second terminal and run the following commands:

.. code-block:: bash

    # Export ROS security environment variables
    export ROS_SECURITY_KEYSTORE=~/deploy_gd_tutorial/keystore
    export ROS_SECURITY_ENABLE=true
    export ROS_SECURITY_STRATEGY=Enforce

    # Source ROS installation and run the talker
    source /opt/ros/${ROS_DISTRO}/setup.bash
    ros2 run demo_nodes_cpp talker --ros-args --enclave /talker_listener/talker

> With the realization of above steps, subsequent output is obtained:

Reviewer: ```suggestion
This should result in the following output:
```

* On host's talker node: ``Publishing: 'Hello World: <number>'``
* While on the remote system side: ``I heard: [Hello World: <number>]``

> 

Reviewer: Please remove the trailing whitespace here.

.. redirect-from::

    Tutorials/Security/Deployment-Guidelines

.. _Deployment-Guidelines:

> 

Reviewer: None of this is necessary, as this is a new guide.

Deployment Guidelines

> ======================

Reviewer: ```suggestion
=====================
```

**Goal:** Understand the best practices when deploying security artifacts into production systems.

**Tutorial level:** Advanced

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

Typical deployment scenarios oftenly involve shipping containerized applications, or packages, into remote systems.
Special attention should be payed when deploying security enabled applications, requiring users to reason about the sensitivity of packaged files.

Complying with the `DDS Security standard <https://www.omg.org/spec/DDS-SECURITY/1.1/About-DDS-SECURITY/>`_,
the ``sros2`` package provides a collection of utilities for managing security under ROS 2 environments in a highly modular and flexible fashion.

> Basic core guidelines on how to organize the different certificates, keys and directories remains to be a critical factor to avoid compromising the security of the system.

Reviewer: ```suggestion
Basic core guidelines on how to organize the different certificates, keys and directories remains a critical factor to avoid compromising the security of the system.
```

This includes protection-awareness and criteria for selecting the minimum set of necessary files to be deployed upon remote production systems for minimizing security exposure.

Prerequisites
-------------

* A docker installation available. Please refer to the installation steps detailed in `Docker installation <https://docs.docker.com/engine/install/>`_.
* (Recommended) A basic understanding on `ROS2 Security design <https://design.ros2.org/articles/ros2_dds_security.html>`_.

> * (Recommended) Previous security tutorials completion.

Reviewer: We should list which ones are recommended here.

General Guidelines
------------------

ROS 2 leverages DDS Security extensions to ensure security on message exchanges within the same enclave.
The different signed files and certificates within an enclave are generated from the private keys and certificates of a `Certificate Authority (CA) <https://en.wikipedia.org/wiki/Certificate_authority>`_ trusted entity.
In fact, two different CA's can be selected for identity and permissions, per enclave.
Those CA artifacts are stored inside ``private/`` and ``public/`` sub-directories of a `Keystore <https://design.ros2.org/articles/ros2_security_enclaves.html>`_ with the following folder structure:

.. code-block:: text

  keystore
  ├── enclaves
  │   └── ...
  │       └── ...
  ├── private
  │   └── ...
  └── public
      └── ...

> A good practice for the creation and usage of a certain Certificate Authority on a typical deployment over a production system, is to:

Reviewer: ```suggestion
A good practice for the creation and usage of a certain Certificate Authority on a typical deployment for a production system, is to:
```

  (1) Create it within the organization system intended for internal use only.
  (2) Generate/modify desired enclaves bearing in mind that:

    * Not all the generated enclaves should be embarked into all target devices.
    * A reasonable way to proceed would be having one enclave per application, allowing for a separation of concerns.

  (3) Ship ``public/`` alongside with corresponding ``enclaves/`` into the different remote production devices during setup.

>   (4) Keep and protect ``private/`` keys and/or certification requests in the organization.

Reviewer: You can use `#.` for the numbering on the list here.  That will automatically renumber.

> It is important to note that if ``private/`` files are lost, it won't be possible to change access permissions, adding or modifying security profiles anymore.

Reviewer: ```suggestion
It is important to note that if ``private/`` files are lost, it won't be possible to change access permissions, add or modify security profiles anymore.
```

* Granting read-only permissions to the ``enclaves/`` directory contents.
* If a PKCS#11 compliant URI is given for generating enclave's private keys, a `Hardware Security Module (HSM) <https://en.wikipedia.org/wiki/Hardware_security_module>`_ could be used to store them.

The following table depicts a summary of the previous statements relating the Keystore directory with the Recommended location:

+------------------------+--------------+---------------+---------------------+
| Directory / Location   | Organization | Target Device | Material Sensitivity|
+========================+==============+===============+=====================+
| public                 |       ✓      |       ✓       |         Low         |
+------------------------+--------------+---------------+---------------------+
| private                |       ✓      |       ✕       |         High        |
+------------------------+--------------+---------------+---------------------+
| enclaves               |       ✓      |       ✓       |        Medium       |
+------------------------+--------------+---------------+---------------------+

Building a deployment scenario
------------------------------

> To illustrate a simple deployment scenario, a docker image will be built intented to be the remote production target device actor.

Reviewer: ```suggestion
To illustrate a simple deployment scenario, a docker image will be built which is intended to be the remote production target device actor.
```

In this example, localhost will serve as the organization's system.
To test security capabilities, a secure listener will be launched on the remote system, whereas a secure talker will be launched on the local host.
Let us start by creating a workspace tree with the following sub-folders:

.. code-block:: bash

  mkdir -p ~/deploy_gd_tutorial/remote_system
  mkdir ~/deploy_gd_tutorial/keystore

Generating a Keystore and necessary Enclaves

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Similarly to previous tutorials, intialize a new keystore tree directory.
This will create *enclaves/* *public/* and *private/* directories, which are explained in more detail in `ROS2 Security enclaves <https://design.ros2.org/articles/ros2_security_enclaves.html>`_.

.. code-block:: bash

  # Source ROS installation
  source /opt/ros/${ROS_DISTRO}/setup.bash
  # Initialize a new keystore directory
  ros2 security create_keystore ~/deploy_gd_tutorial/keystore

Next, create an enclave for the local talker node within the */keystore* directory.

.. code-block:: bash

  # Create secure talker's enclave
  ros2 security create_enclave ~/deploy_gd_tutorial/keystore /talker_listener/talker

At this point, step into the remote_system workspace, create the corresponding enclave and copy just the *public/* and *enclaves/* directories to the current one.
Those security artifacts will be needed by the remote system to enable listener's security.

> For the sake of simplicity, the same CA is used within this enclave for both, identity and permissions.

Reviewer: ```suggestion
For the sake of simplicity, the same CA is used within this enclave for both identity and permissions.
```

> Note that *private/* folder is not moved but left in local host (organization).

Reviewer: ```suggestion
Note that *private/* folder is not moved but left in localhost (organization).
```

.. code-block:: bash

  # Create an enclave for the secure listener's enclave
  ros2 security create_enclave ~/deploy_gd_tutorial/keystore /talker_listener/listener

At the end of these steps, the structure of */enclaves* sub-directory within *~/deploy_gd_tutorial/keystore* should look like the following:

.. code-block:: text

  enclaves
  ├── governance.p7s
  ├── governance.xml
  └── talker_listener
      ├── listener
      │   ├── cert.pem
      │   ├── governance.p7s
      │   ├── identity_ca.cert.pem
      │   ├── key.pem
      │   ├── permissions_ca.cert.pem
      │   ├── permissions.p7s
      │   └── permissions.xml
      └── talker
          ├── cert.pem
          ├── governance.p7s
          ├── identity_ca.cert.pem
          ├── key.pem
          ├── permissions_ca.cert.pem
          ├── permissions.p7s
          └── permissions.xml

> Now, create and populate the */keystore* directory that will be embarked onto the remote system with necessary files only.

Reviewer: ```suggestion
Now, create and populate the */keystore* directory that will be copied onto the remote system with only the necessary files.
```

.. code-block:: bash

  # Move to remote system path
  cd ~/deploy_gd_tutorial/remote_system
  # Ship governance files, listener enclave and public/ directories only
  # minimizing security threat
  mkdir -p keystore/enclaves/talker_listener
  cp -R ../keystore/public keystore
  cp -R ../keystore/enclaves/governance.* keystore/enclaves
  cp -R ../keystore/enclaves/talker_listener/listener keystore/enclaves/talker_listener

> Introducing the former commands, the current ``~/deploy_gd_tutorial/remote_system`` directory should be:

Reviewer: ```suggestion
After the former commands, the current ``~/deploy_gd_tutorial/remote_system`` directory should be:
```

.. code-block:: text

  remote_system
  └── keystore
      ├── enclaves
      │   ├── governance.p7s
      │   ├── governance.xml
      │   └── talker_listener
      │       └── listener
      │           ├── cert.pem
      │           ├── governance.p7s
      │           ├── identity_ca.cert.pem
      │           ├── key.pem
      │           ├── permissions_ca.cert.pem
      │           ├── permissions.p7s
      │           └── permissions.xml
      └── public
          ├── ca.cert.pem
          ├── identity_ca.cert.pem
          └── permissions_ca.cert.pem

Creating remote's system docker image

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> To get started, head into remotes's workspace path with:

Reviewer: ```suggestion
To get started, change into the remotes's workspace path with:
```

Create a new file ``Dockerfile`` in the same directory with preferred text editor.

.. code-block:: bash

  ARG ROS_DISTRO=humble
  FROM ros:${ROS_DISTRO}-ros-base

  RUN apt-get update && apt-get install -y \
        ros-${ROS_DISTRO}-demo-nodes-cpp \
        ros-${ROS_DISTRO}-demo-nodes-py && \
      rm -rf /var/lib/apt/lists/*

  ARG KEYSTORE_DIR=/keystore

  RUN mkdir -p ${KEYSTORE_DIR}/enclaves \
    mkdir ${KEYSTORE_DIR}/public

  COPY keystore ${KEYSTORE_DIR}

  ENV ROS_SECURITY_KEYSTORE=${KEYSTORE_DIR}
  ENV ROS_SECURITY_ENABLE=true
  ENV ROS_SECURITY_STRATEGY=Enforce

  COPY entrypoint.sh /entrypoint.sh
  RUN chmod +x /entrypoint.sh

  ENTRYPOINT ["/entrypoint.sh"]

The resultant directory structure should be the one depicted below:

.. code-block:: text

  remote_system
    ├── Dockerfile
    ├── entrypoint.sh
    └── keystore
        ├── enclaves
        │   ├── ...
        └── public
            ├── ...

Build the docker image with the command:

.. code-block:: bash

  # Build remote's system image
  docker build -t ros2_security/deployment_tutorial .

Running the example
-------------------

Launch the following commands in two different terminals:

.. tabs::

>     .. tab:: TERMINAL 1

Reviewer: We haven't typically used tabs for multiple terminals (see for example https://docs.ros.org/en/rolling/Tutorials/Advanced/Recording-A-Bag-From-Your-Own-Node-CPP.html#build-and-run). So just have inline text there that says to open terminal 1, then terminal 2.

        .. code-block:: bash

            # Start remote system container
            docker run -it ros2_security/deployment_tutorial

    .. tab:: TERMINAL 2

        .. code-block:: bash

            # Export ROS security environment variables
            export ROS_SECURITY_KEYSTORE=~/deploy_gd_tutorial/keystore
            export ROS_SECURITY_ENABLE=true
            export ROS_SECURITY_STRATEGY=Enforce

            # Source ROS installation and run the talker
            source /opt/ros/${ROS_DISTRO}/setup.bash
            ros2 run demo_nodes_cpp talker --ros-args --enclave /talker_listener/talker

With the realization of above steps, the subsequent output is obtained:

.. raw:: html

    <video width=100% height=auto autoplay loop controls muted>
        <source src="../../../_static/security_deployment_tutorial.mp4">
        Your browser does not support the video tag.
    </video>

> 

Reviewer: Please remove this and the video. We don't currently host videos, though we are in the process of coming up with a plan for showing community videos.
